### DISCUSSION TITLE
Motion Gen result success is True but robot is still collides

### DISCUSSION THREAD

**Discussion by vignesh-anand** [NVlabs/curobo#285] [Software Q&A]
Hi Curobo Team,
We are using Curobo to plan for the panda mobile, with simulation in MuJoCo DM Control. Our collision world is shown below
![image](https://github.com/NVlabs/curobo/assets/56486572/159c2fea-aa65-43ac-8b45-7e03a789e142)

`physics.reset(0)
update_curobo_world(physics,world_config_panda1,'panda1/')
motion_gen_config_1 = MotionGenConfig.load_from_robot_config(
    "franka_mobile.yml",
    world_config_panda1,
    interpolation_dt=0.01,
    collision_checker_type=CollisionCheckerType.MESH,collision_activation_distance=0.01,trajopt_dt=0.05
)
motion_gen1 = MotionGen(motion_gen_config_1)
motion_gen1.warmup(parallel_finetune=True)

#retract_cfg = motion_gen.get_retract_config()
print("warm")
update_curobo_world(physics,world_config_panda1,'panda1/')

update_curobo_world(physics,world_config_panda1,'panda1/')
world_config_class=WorldConfig.from_dict(world_config_panda1)
motion_gen1.update_world(world_config_class)
goal_pose = Pose.from_list([0.7, 0.5, 0.5, 0.0, 1.0, 0.0, 0.0],q_xyzw=False,tensor_args=TensorDeviceType(dtype=torch.float32))  # x, y, z, qw, qx, qy, qz
start_state = JointState.from_position(
    torch.tensor(physics.data.qpos[[0,1,2,3,4,5,6,7]],dtype=torch.float32).view(1,-1).cuda(),
    #joint_names=[
        #"shoulder_pan_joint",
        #"shoulder_lift_joint",
        #"elbow_joint",
        #"wrist_1_joint",
        #"wrist_2_joint",
        #"wrist_3_joint",
    #],
)

result = motion_gen1.plan_single(start_state, goal_pose, MotionGenPlanConfig(max_attempts=2,parallel_finetune=True,enable_graph=True,enable_opt=True))
traj = result.get_interpolated_plan()  # result.optimized_dt has the dt between timesteps
print("Trajectory Generated: ", result.success)`

We get result.success as True, but when we execute the trajectory it appears to result in a collision

https://github.com/NVlabs/curobo/assets/56486572/3a8082a9-3fe3-46f3-a2dc-6ce915047177

Is this a parameter-tuning problem? We are using graph planning to seed the optimizer. We found a mention of speed_metric but cannot seem to find how to use this. Activation distance helps slightly, but beyond 0.1, the planner fails to converge. For poses where the two arms are separated, the planner succeeds.

**Comment by vignesh-anand** [comment]
https://github.com/NVlabs/curobo/assets/56486572/faa594ec-c68e-4265-a9d5-3dfb78603db2

Planning with the graph planner alone does result in a collision-free trajectory, but with a very small step size, which when executed results in a very slow movement. We are not sure how to control this step size.

**Comment by balakumar-s** [comment]
This looks like a tracking problem where the low-level controller in sim is unable to track the trajectory generated by cuRobo. You can reduce the speed of the trajectory being planned by adding `time_dilation_factor=0.1` when planning:

Change your code to below:

```
result = motion_gen1.plan_single(start_state, goal_pose, MotionGenPlanConfig(max_attempts=2,time_dilation_factor=0.1))
traj = result.get_interpolated_plan() 
print("Trajectory Generated: ", result.success)`
```


API: https://curobo.org/_api/curobo.wrap.reacher.motion_gen.html#curobo.wrap.reacher.motion_gen.MotionGenPlanConfig.time_dilation_factor

**Comment by vignesh-anand** [comment]
Hi Bala,
Yup, I found the tracking problem soon after posting this. Do you have advice on optimizing the plan time? My current implementation above is taking about 3 seconds on The GTX 1650 on my laptop. Is this expected or this slow? The website says planning on the order of ms?

**Reply by balakumar-s** [reply]
Can you run `python benchmark/curobo_benchmark.py` to see what numbers you are getting? We provide numbers for 4090 here: https://curobo.org/get_started/4_benchmarks.html#latest-motion-generation-results

**Reply by vignesh-anand** [reply]
Hi Bala,
Here are my GPU stats and the first few tests from the benchmark. Do you thunk I just have slow hardware or can I optimize my implementation and use further?

My current times on my use case:  Motion Gen with Graph+traj_opt (Mesh): 2.5-3 seconds)
Motion Gen (Graph+traj_opt (Primitive): 0.4 seconds
create_obbworld: 0.4. 

A picture of my collision world is in a post above.

Appreciate the Support!

+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 550.54.15              Driver Version: 550.54.15      CUDA Version: 12.4     |
|-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  NVIDIA GeForce GTX 1650 ...    Off |   00000000:01:00.0 Off |                  N/A |
| N/A   61C    P0             26W /   30W |    2791MiB /   4096MiB |     79%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+------------------------+----------------------+
                                                                                         
+-----------------------------------------------------------------------------------------+
| Processes:                                                                              |
|  GPU   GI   CI        PID   Type   Process name                              GPU Memory |
|        ID   ID                                                               Usage      |
|=========================================================================================|
|    0   N/A  N/A      1035      G   /usr/lib/xorg/Xorg                              4MiB |
|    0   N/A  N/A     45817      C   ...esh/miniconda3/envs/roco/bin/python       1970MiB |
|    0   N/A  N/A     49397      C   python                                        814MiB |


*****RUN: 0
  0%|                                                                          | 0/8 [00:00<?, ?it/sbookshelf_small_panda 100.00 0.12211875438690185 0.47823251247405973 0.24446555453323526 2.527694068402052 5.128851804733276 2.0619282436370847
mean: 1.120 ± 0.621 median:1.000 75%: 1.000 98%: 3.020
 12%|████████▎                                                         | 1/8 [00:22<02:39, 22.73s/itbookshelf_tall_panda 100.00 0.13153625249862672 0.2806796169281005 0.25275991058151703 2.1799578719679267 4.364956445693969 1.4900767421722412
mean: 1.200 ± 1.503 median:1.000 75%: 1.000 98%: 2.000
 25%|████████████████▌                                                 | 2/8 [00:50<02:35, 25.84s/itbookshelf_thin_panda 100.00 0.10383365869522095 0.21438111782073974 0.22843045542686014 2.0544340655207636 4.405650587081909 1.617396821975708
mean: 1.010 ± 0.099 median:1.000 75%: 1.000 98%: 1.000
 38%|████████████████████████▊                                         | 3/8 [01:15<02:07, 25.53s/itbox_panda 100.00 0.10694597721099854 0.4244799757003783 0.09102236510898365 0.6481795062404125 4.16153016090393 1.6345809912681577
mean: 1.050 ± 0.218 median:1.000 75%: 1.000 98%: 2.000
 50%|█████████████████████████████████                                 | 4/8 [01:30<01:25, 21.39s/itbox_panda_flipped 100.00 0.11257483005523682 0.2253070592880247 0.037988317405961425 0.6442624783888459 4.82073554039001 2.007324278354642
mean: 1.160 ± 1.398 median:1.000 75%: 1.000 98%: 2.000
 62%|█████████████████████████████████████████▎                        | 5/8 [01:47<00:58, 19.58s/itcage_panda 100.00 0.10648121118545532 0.23650248050689587 0.2456785840786324 2.2261693823337554 5.368900356292724 1.5370596599578856
mean: 1.030 ± 0.171 median:1.000 75%: 1.000 98%: 2.000
 75%|█████████████████████████████████████████████████▌                | 6/8 [02:01<00:35, 17.65s/ittable_pick_panda 100.00 0.10105192184448242 0.1948185586929319 0.2675583836844453 2.129366885274649 4.656328563690185 1.3267771220207214
mean: 1.010 ± 0.099 median:1.000 75%: 1.000 98%: 1.000
 88%|█████████████████████████████████████████████████████████▊        | 7/8 [02:17<00:17, 17.05s/ittable_under_pick_panda 100.00 0.2426144185692373 0.7724231815338126 0.09271786106495196 0.6911638452285769 6.755357742309566 2.698440227508543
mean: 1.929 ± 0.879 median:2.000 75%: 2.000 98%: 4.040
100%|██████████████████████████████████████████████████████████████████| 8/8 [02:47<00:00, 20.89s/it]
All:  100.00 2.2299595642089844 0.12825167462584075 0.09924089908599854 0.19252825586590916 2.8112598657608032
  0%|                                                                         | 0/12 [00:00<?, ?it/scubby_neutral_goal 100.00 0.08749325533170958 0.09608819484710693 0.23036982146722212 0.9784925663592042 4.9905535793304425 1.4954320764541624
mean: 1.000 ± 0.000 median:1.000 75%: 1.000 98%: 1.000
  8%|█████▍                                                           | 1/12 [00:09<01:43,  9.42s/itcubby_neutral_start 100.00 0.08671086947123209 0.1322080612182617 0.6015985579748909 2.1903751798470816 4.512556037902832 1.3215572834014893
mean: 1.013 ± 0.115 median:1.000 75%: 1.000 98%: 1.000
 17%|██████████▋

**Reply by balakumar-s** [reply]
The benchmark is giving you 100ms planning time while your custom script is giving 400ms. We can definitely improve. Looking at your script, you can change two things:
1. Change `trajopt_dt=0.25` which is the default value. This should help in converging with 1 attempt.
```
motion_gen_config_1 = MotionGenConfig.load_from_robot_config(
"franka_mobile.yml",
world_config_panda1,
interpolation_dt=0.01,
collision_checker_type=CollisionCheckerType.MESH,collision_activation_distance=0.01,trajopt_dt=0.05
)
```
2. Change your MotionGenPlanConfig to MotionGenPlanConfig(max_attempts=10). Don't change any of the other values.

Are you running any other tasks on the GPU?

**Reply by frankchan12138** [reply]
hello Balakumar Sundaralingam! I try cuMotion moveit plugin which use cuRobo as the planner. And my compute platform is **Jetson AGX Orin 64GB（at the MAXN mode）**. I found that when I call `plan_single` to generate motion in a plan scene without obstacles,  It takes about **300ms ~ 500ms**. While the paper says that the mean plan time is about 100ms. I open the debug mode of curobo and get the following log(the timestamp is added by my self). Is there something wrong with my config? I have already change `trajopt_dt` to **0.25** and `max_attempts` to **10**. Any advice on this problem will be highly appreciated, Thanks

> [INFO] [1726756133.839344250] [cumotion_action_server]: before plan_single
[INFO] [1726756133.842644137] [cumotion_action_server]: mg_plan_config is MotionGenPlanConfig(enable_graph=False, enable_opt=True, use_nn_ik_seed=False, need_graph_success=False, max_attempts=10, timeout=10.0, enable_graph_attempt=3, disable_graph_attempt=None, ik_fail_return=None, partial_ik_opt=False, num_ik_seeds=None, num_graph_seeds=None, num_trajopt_seeds=None, success_ratio=1, fail_on_invalid_query=True, use_start_state_as_retract=True, pose_cost_metric=None, enable_finetune_trajopt=True, parallel_finetune=True, finetune_dt_scale=0.85, finetune_attempts=5, finetune_dt_decay=1.025, time_dilation_factor=0.1, check_start_validity=True, finetune_js_dt_scale=1.1)
2024-09-19 22:28:53,842 842  - curobo - INFO - Planning for Single Goal: 1
2024-09-19 22:28:53,843 843  - curobo - INFO - num_ik_seeds 4 num_trajopt_seeds 2 num_graph_seeds 2
2024-09-19 22:28:53,867 867  - curobo - INFO - MG Iter: 0
2024-09-19 22:28:53,881 881  - curobo - INFO - Updating safety params
2024-09-19 22:28:53,884 884  - curobo - INFO - Updating optimizer params
2024-09-19 22:28:53,896 896  - curobo - INFO - current opt use_cuda_graph flag is True
2024-09-19 22:28:53,898 898  - curobo - INFO - current opt use_cuda_graph flag is True
2024-09-19 22:28:53,917 917  - curobo - INFO - self.opt_dt is 0.020124197006225586
2024-09-19 22:28:53,943 943  - curobo - INFO - MG: running TO
2024-09-19 22:28:53,945 945  - curobo - INFO - TrajOpt: solving with Pose batch:1
2024-09-19 22:28:53,958 958  - curobo - INFO - Updating safety params
2024-09-19 22:28:53,959 959  - curobo - INFO - Updating optimizer params
2024-09-19 22:28:53,960 960  - curobo - INFO - current opt use_cuda_graph flag is True
2024-09-19 22:28:53,961 961  - curobo - INFO - current opt use_cuda_graph flag is True
2024-09-19 22:28:53,985 985  - curobo - INFO - self.opt_dt is 0.02456831932067871
2024-09-19 22:28:53,989 989  - curobo - INFO - Ran TO
2024-09-19 22:28:54,020 020  - curobo - INFO - TrajOpt: solving with Pose batch:1
2024-09-19 22:28:54,029 029  - curobo - INFO - Updating safety params
2024-09-19 22:28:54,036 036  - curobo - INFO - Updating optimizer params
2024-09-19 22:28:54,050 050  - curobo - INFO - current opt use_cuda_graph flag is True
2024-09-19 22:28:54,136 136  - curobo - INFO - self.opt_dt is 0.08515596389770508
2024-09-19 22:28:54,146 146  - curobo - INFO - Ran TO
[INFO] [1726756134.207989340] [cumotion_action_server]: after plan_single

In addition, the following image is the jtop screenshot of Orin, in which cumotion_planner is the cuMotion plugin using cuRobo as planner
![2024-09-19_22-46](https://github.com/user-attachments/assets/7c8abab1-6000-4bbd-a2f5-e62ae1373c74)
